include/toolchain/common.h

```diff
@@ -144,4 +144,48 @@
 #define BUILD_ASSERT_MSG(EXPR, MSG) BUILD_ASSERT(EXPR)
 #endif
 
+/*
+ * This is meant to be used in conjonction with __in_section() and similar
+ * where scattered structure instances are concatened together by the linker
+ * and walked by the code at run time just like a contiguous array of such
+ * structures.
+ *
+ * Assemblers and linkers may insert alignment padding by default whose
+ * size is larger than the natural alignment for those structures when
+ * gathering various section segments together, messing up the array walk.
+ * To prevent this, we need to provide an explicit alignment not to rely
+ * on the default that might just work by luck.
+ *
+ * Alignment statements in  linker scripts are not sufficient as
+ * the assembler may add padding by itself to each segment when switching
+ * between sections within the same file even if it merges many such segments
+ * into a single section in the end.
+ */
+#define __decl_align(type) __aligned(__alignof(type)) type
```

Comment:
As mentioned by someone else, MISRA forbids us from introducing new symbols with a prepended underscore.  Should be "Z_DECL_ALIGN" by current convention.